jni-gen/src/wrapper_generator.rs

```diff
@@ -62,6 +66,9 @@ impl WrapperGenerator {
 
         //remove_dir_all(out_dir)?;
         create_dir_all(out_dir)?;
+        create_dir_all(out_dir.join("builtins"))?;
+        let builtins_utils = Path::new("/home/simon/repo/prusti-dev/jni-gen/builtins/utils.rs");
```

Comment:
This will work only on my machine. Is there an elegant way to get a path to the current file? I could come up with some working solution (like taking the path to the current crate, go up until you find the root directory of the repository (e.g. by cargo.toml having workspace) and then adding the relative path from the root of the repository, but I guess that's not a very robust/clean solution.

Reply:
See this to obtain the path of the current module: https://doc.rust-lang.org/std/macro.module_path.html. Then, you can join `../builtins`. For maintainability of `jni-gen`, it's better to never make assumptions about where the crate is placed in the repository or how Prusti uses it.

Reply:
I checked the `module_path!` macro, but it returns the path in terms of modules - `jni_gen::wrapper_generator`
I also tried the `file!` macro, but it only returns the path to the current file within the crate (`jni-gen/src/wrapper_generator.rs`).
However, I am unable to obtain the absolute path to the `jni_gen` crate.
I can get the absolute path of the crate that is being built by `std::env::current_dir()` (which should be the same as `env::var("CARGO_MANIFEST_DIR")`).
This can be `/home/simon/repo/prusti-dev/viper-sys` or `/home/simon/repo/prusti-dev/jni-gen/systest` (the things before `jni-gen` are just for my PC).

but I don't know how to get the path to the `jni-gen` crate. I (incorrectly, as you stated) assume that the `jni-gen` is in the root of the repository and get the repository root by truncating the `current_dir()` path.

Reply:
Instead of *copying* the file using `jni-gen`, you can compile the file *into* `jni-gen` with something like [`include_str!`](https://doc.rust-lang.org/std/macro.include_str.html) and just make it write the file.

Reply:
> I checked the module_path! macro, but it returns the path in terms of modules - jni_gen::wrapper_generator

Oh, my bad.

> I also tried the file! macro, but it only returns the path to the current file within the crate (jni-gen/src/wrapper_generator.rs).

That sounds good, right? Can `file!()` joined with `../../builtins` be used as path of the source folder of the copy?

Reply:
> That sounds good, right? Can `file!()` joined with `../../builtins` be used as path of the source folder of the copy?

That gives a relative path within the `jni-gen`, but when the file is copied it is during the build (running build.rs) of either `systest` or `viper-sys`, but I would then need to combine it with path to `jni-gen`, which I don't have (basically the path that is written in the `cargo.toml` files as a path to `jni-gen`).

I think the @Aurel300 approach might be an easy fix for this.

Reply:
Oh, I see. Yes, Aurel's approach sounds good. Alternatively, to obtain the path of the crate being compiled you can use `std::env!("CARGO_MANIFEST_DIR")`, which evaluates at *compilation time* the environment variable set by cargo.

Reply:
The PR uses `include_str!("../builtins/mod.rs");`, so we can close this thread.